Component Based `Mounter` and `Receiver` types.

Samyoul opened the pull request:
## What's Changed?

The main focus of this PR is to simplify the implementation of `Mounter` and `Receiver` types. This PR introduces the following:

- `BasePayloadMounter`
- `BasePayloadReceiver`

Both of these new types **replace** the respective `PayloadMounter` and `PayloadReceiver` implementations:

- **`PayloadMounter` :**
  - `AccountPayloadMounter`
  - `RawMessagePayloadMounter`
  - `InstallationPayloadMounter`

- **`PayloadReceiver` :**
  - `AccountPayloadReceiver`
  - `RawMessagePayloadReceiver`
  - `InstallationPayloadReceiver`
  
The `Mounter` and `Receiver` types all followed the same process flow for `Mount()` and `Receive()` and because these types have `Loader` and `Storer` interface fields types, the constituent components are interchangable.

### `Mounter`

```go
type BasePayloadMounter struct {
	*PayloadLockPayload
	*PayloadToSend

	loader     PayloadLoader
	marshaller ProtobufMarshaller
	encryptor  *PayloadEncryptor
}
```

The `Mount()` function

```go
// Mount loads and prepares the payload to be stored in the PayloadLoader's state ready for later access
func (bpm *BasePayloadMounter) Mount() error {
	err := bpm.loader.Load()
	if err != nil {
		return err
	}

	p, err := bpm.marshaller.MarshalProtobuf()
	if err != nil {
		return err
	}

	return bpm.encryptor.encrypt(p)
}
```

### `Receiver`

```go
type BasePayloadReceiver struct {
	*PayloadLockPayload
	*PayloadReceived

	encryptor    *PayloadEncryptor
	unmarshaller ProtobufUnmarshaller
	storer       PayloadStorer

	receiveCallback func()
}
```

The `Receive()` function

```go
// Receive takes a []byte representing raw data, parses and stores the data
func (bpr *BasePayloadReceiver) Receive(data []byte) error {
	err := bpr.encryptor.decrypt(data)
	if err != nil {
		return err
	}

	err = bpr.unmarshaller.UnmarshalProtobuf(bpr.Received())
	if err != nil {
		return err
	}

	if bpr.receiveCallback != nil {
		bpr.receiveCallback()
	}

	return bpr.storer.Store()
}
```

### Type instantiation Example:

All the replaced types are now defined by a `Base` type's components and not a dedicated type. Example:

```go
// NewAccountPayloadMounter generates a new and initialised AccountPayload flavoured BasePayloadMounter
// responsible for the whole lifecycle of an AccountPayload
func NewAccountPayloadMounter(pe *PayloadEncryptor, config *SenderConfig, logger *zap.Logger) (*BasePayloadMounter, error) {
	l := logger.Named("AccountPayloadLoader")
	l.Debug("fired", zap.Any("config", config))

	pe = pe.Renew()

	// A new SHARED AccountPayload
	p := new(AccountPayload)
	apl, err := NewAccountPayloadLoader(p, config)
	if err != nil {
		return nil, err
	}

	return NewBasePayloadMounter(
		apl,
		NewPairingPayloadMarshaller(p, l),
		pe,
	), nil
}
```

Note the instantiation of a `AccountPayloadMounter` is basically unchanged, apart from passing the 3 main components of a `Mounter` to a `BasePayloadMounter` instead of a dedicated `AccountPayloadMounter`. This allows for better code reusability and component interchangeability.

---

### Facilitation

To facilitate this simplification I've implemented / restructured the following types:

- `PayloadEncryptor` functionality / embeddables
  - `PayloadLockPayload`
  - `PayloadToSend`
  - `PayloadReceived`
- `RawMessagesPayload`
- `RawMessagePayloadMarshaller`

The `PayloadEncryptor` embeddable types allow any struct to embed specific API facing functions, allowing the embedding struct to implement a subset of `PayloadMounter` or `PayloadReceiver` method signatures that make use of unexported `PayloadEncryptor` functionality.

#### Example:

```go
type BasePayloadMounter struct {
	*PayloadLockPayload
	*PayloadToSend

	loader     PayloadLoader
	marshaller ProtobufMarshaller
	encryptor  *PayloadEncryptor
}
```

Instantiated via:

```go
func NewBasePayloadMounter(loader PayloadLoader, marshaller ProtobufMarshaller, e *PayloadEncryptor) *BasePayloadMounter {
	return &BasePayloadMounter{
		PayloadLockPayload: &PayloadLockPayload{e},
		PayloadToSend:      &PayloadToSend{e},
		loader:             loader,
		marshaller:         marshaller,
		encryptor:          e,
	}
}
```

Notice the embeddable `PayloadEncryptor` types:

```go
type BasePayloadMounter struct {
	*PayloadLockPayload
	*PayloadToSend
}
```

Which gives the `BasePayloadMounter` the following methods:

```go
func (pl *PayloadLockPayload) LockPayload() {
	pl.lockPayload()
}
```

```go
func (pts *PayloadToSend) ToSend() []byte {
	return pts.getEncrypted()
}
```

Fulfilling the `LockPayload()` and `ToSend()` method requirements of the `PayloadMounter` interface.

### `RawMessagePayloadMarshaller`

`RawMessagesPayload` and `RawMessagePayloadMarshaller` were implemented for a few reasons;

- Segregate RawMessage data marshalling into a dedicated type, separating loading, storing and transport concerns.
- To allow the `BasePayloadMounter` and `BasePayloadReceiver` to take advantage of a `Loader`, `Storer`, `Encryptor` **and** **`Marshaller`** functionality in their respective `Mount()` and `Receive()` functions.

---

I've also added a little fix for struct extending where the incoming struct does not have a value set for its `publicKey` value.

And I've added a little utility function to check if the `Settings` struct is empty.

## Why Change?

To improve code maintainability, and all the reasons I gave in the above section 😆 .

Closes https://github.com/status-im/status-go/issues/3302

Samyoul committed:
Rolled LockPayload, ToSend and Received funcs into dedicated types

Samyoul committed:
Implemented structure of BasePayloadMounter.Mount()

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [ ] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://status.im/docs/))?
- [ ] Have you tested changes with mobile?
- [ ] Have you tested changes with desktop?

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (28)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 5e3c360d | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3329/1/) | 2023-03-22 14:08:50 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230322-140609-5e3c36-pr3329.zip) |
| :heavy_check_mark: | 5e3c360d | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3329/1/) | 2023-03-22 14:09:47 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230322-140605-5e3c36-pr3329.zip) |
| :heavy_multiplication_x: | 5e3c360d | [#1](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3329/1/) | 2023-03-22 14:10:41 | ~4 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3329/1//consoleText) |
| :heavy_check_mark: | 5e3c360d | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3329/1/) | 2023-03-22 14:12:13 | ~6 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230322-140604-5e3c36-pr3329.aar) |
| :heavy_check_mark: | 5e3c360d | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3329/2/) | 2023-03-24 17:55:20 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230324-175349-5e3c36-pr3329.zip) |
| :heavy_multiplication_x: | 5e3c360d | [#2](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3329/2/) | 2023-03-24 17:55:52 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3329/2//consoleText) |
| :heavy_check_mark: | 5e3c360d | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3329/2/) | 2023-03-24 17:56:25 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230324-175349-5e3c36-pr3329.zip) |
| :heavy_check_mark: | 5e3c360d | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3329/2/) | 2023-03-24 17:58:10 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230324-175345-5e3c36-pr3329.aar) |
| | | | | | | |
| :heavy_check_mark: | cd9bcaee | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3329/3/) | 2023-03-28 12:20:21 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230328-121905-cd9bca-pr3329.zip) |
| :heavy_multiplication_x: | cd9bcaee | [#3](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3329/3/) | 2023-03-28 12:21:13 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3329/3//consoleText) |
| :heavy_check_mark: | cd9bcaee | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3329/3/) | 2023-03-28 12:22:51 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230328-121901-cd9bca-pr3329.aar) |
| :heavy_check_mark: | cd9bcaee | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3329/3/) | 2023-03-28 12:28:23 | ~9 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230328-121905-cd9bca-pr3329.zip) |
| | | | | | | |
| :heavy_check_mark: | 5c660293 | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3329/4/) | 2023-03-29 17:14:12 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230329-171302-5c6602-pr3329.zip) |
| :heavy_check_mark: | 5c660293 | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3329/4/) | 2023-03-29 17:17:33 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230329-171302-5c6602-pr3329.aar) |
| :heavy_check_mark: | 5c660293 | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3329/4/) | 2023-03-29 17:22:43 | ~9 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230329-171302-5c6602-pr3329.zip) |
| :heavy_check_mark: | 5c660293 | [#4](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3329/4/) | 2023-03-29 17:23:20 | ~10 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3329/4//consoleText) |
| | | | | | | |
| :heavy_check_mark: | ff4abf3a | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3329/5/) | 2023-03-30 10:02:29 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230330-100106-ff4abf-pr3329.zip) |
| :heavy_check_mark: | ff4abf3a | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3329/5/) | 2023-03-30 10:05:31 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230330-100106-ff4abf-pr3329.aar) |
| :heavy_check_mark: | ff4abf3a | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3329/5/) | 2023-03-30 10:10:42 | ~9 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230330-100106-ff4abf-pr3329.zip) |
| :heavy_check_mark: | ff4abf3a | [#5](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3329/5/) | 2023-03-30 10:13:09 | ~12 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3329/5//consoleText) |
| | | | | | | |
| :heavy_check_mark: | e2d616ea | [#6](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3329/6/) | 2023-04-02 23:10:07 | ~58 sec | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230402-230908-e2d616-pr3329.zip) |
| :heavy_check_mark: | e2d616ea | [#6](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3329/6/) | 2023-04-02 23:12:35 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230402-230908-e2d616-pr3329.aar) |
| :heavy_check_mark: | e2d616ea | [#6](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3329/6/) | 2023-04-02 23:14:29 | ~5 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230402-230908-e2d616-pr3329.zip) |
| :heavy_check_mark: | e2d616ea | [#6](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3329/6/) | 2023-04-02 23:20:57 | ~11 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3329/6//consoleText) |
| | | | | | | |
| :heavy_check_mark: | f39c1324 | [#7](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3329/7/) | 2023-04-03 13:52:19 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230403-134928-f39c13-pr3329.zip) |
| :heavy_check_mark: | f39c1324 | [#7](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3329/7/) | 2023-04-03 13:54:03 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230403-134925-f39c13-pr3329.zip) |
| :heavy_check_mark: | f39c1324 | [#7](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3329/7/) | 2023-04-03 13:55:13 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230403-134925-f39c13-pr3329.aar) |
| :heavy_multiplication_x: | f39c1324 | [#7](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3329/7/) | 2023-04-03 14:13:57 | ~24 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3329/7//consoleText) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | a8485a3c | [#8](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3329/8/) | 2023-04-03 13:58:52 | ~53 sec | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230403-135758-a8485a-pr3329.zip) |
| :heavy_check_mark: | a8485a3c | [#8](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3329/8/) | 2023-04-03 14:01:06 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230403-135758-a8485a-pr3329.zip) |
| :heavy_check_mark: | a8485a3c | [#8](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3329/8/) | 2023-04-03 14:02:48 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230403-135758-a8485a-pr3329.aar) |
| :heavy_check_mark: | a8485a3c | [#8](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3329/8/) | 2023-04-03 14:27:30 | ~13 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3329/8//consoleText) |
| | | | | | | |
| :heavy_check_mark: | c06a6605 | [#9](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3329/9/) | 2023-04-04 10:29:07 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230404-102801-c06a66-pr3329.zip) |
| :heavy_check_mark: | c06a6605 | [#9](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3329/9/) | 2023-04-04 10:30:41 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230404-102801-c06a66-pr3329.zip) |
| :heavy_check_mark: | c06a6605 | [#9](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3329/9/) | 2023-04-04 10:31:42 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230404-102801-c06a66-pr3329.aar) |
| :heavy_check_mark: | c06a6605 | [#9](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3329/9/) | 2023-04-04 10:39:18 | ~11 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3329/9//consoleText) |

Samyoul committed:
Implemented base Mount() func

This uses Load, MarshalProtobuf and encrypt process flow for any PayloadMounter type. In fact the concept of a PayloadMounter type is only useful now in the instantiation context.

Samyoul committed:
Removed redudant PayloadMounter types

Samyoul committed:
Refactored BasePayloadReceiver to handle Receive()

Additionally to allow this process flow I refactored RawMessagePayloadReceiver and InstallationPayloadHandler to use a dedicated Marshaller type. Also added a fix to struct extention functionality, we want to ignore the process if there is no public key because that will key encoding. Seems an unnecessary bug to have to handle when you know there is no key.

Samyoul commented on server/pairing/interfaces.go:
```diff
@@ -25,14 +25,17 @@ type PayloadLocker interface {
 	LockPayload()
 }
 
+// TODO if this interface only gets a logger, then maybe remove the interface and change consuming function params
+//  to accept a *zap.logger
+
 type HandlerServer interface {
 	GetLogger() *zap.Logger
 }
```
I'll make an issue for this and update the code with a link to the issue.

https://github.com/status-im/status-go/issues/3370

ilmotta reviewed (approved):
Nothing to add, and nice round of refactoring!

qfrank commented on server/pairing/raw_message_collector.go:
```diff
@@ -21,13 +21,13 @@ func (r *RawMessageCollector) getRawMessages() []*common.RawMessage {
 }
 
-func (r *RawMessageCollector) convertToSyncRawMessage() *protobuf.SyncRawMessage {
+func (r *RawMessageCollector) convertToSyncRawMessage() protobuf.SyncRawMessage {
```
can you explain why not use pointer , thank you 🙂

Samyoul replied:
Ah, actually this doesn't need to be a value now. I previously during this refactor had code that passed out a value and assigned to a pointer's value. But during this refactor, I realised that `protobuf.SyncRawMessage` is the transport wrapper and not the actual payload, so I changed the refactor to use a `RawMessagePayload` as the payload for loading, storing and marshalling.

So I will change this back to a pointer, because it is completely fine for it to be a pointer.

qfrank commented on server/pairing/raw_message_handler.go:
```diff
@@ -38,18 +35,19 @@ func (s *SyncRawMessageHandler) CollectInstallationData(rawMessageCollector *Raw
 
-func (s *SyncRawMessageHandler) PrepareRawMessage(keyUID, deviceType string) ([]byte, error) {
+func (s *SyncRawMessageHandler) PrepareRawMessage(keyUID, deviceType string) (rm []*protobuf.RawMessage, as []*accounts.Account, ss *settings.Settings, err error) {
+	ss = new(settings.Settings)
```
Regarding the abbreviation of variable names, I'd like to share my thoughts. If the abbreviated variable is only used in a small scope, such as within 5 lines of code, I can accept it. However, if it is used within a 50-line function, I believe it increases the complexity of reading the code. 🙂

Samyoul replied:
@qfrank So normally I wouldn't change a variable name in existing code, however in this case I changed the variable because the name `settings` collides with / shadows the package `github.com/status-im/status-go/multiaccounts/settings`. So I changed the name to the next most idiomatic name I could think of.

I am very happy to consider any suggestions.

qfrank replied:
My previous expression may not have been clear enough. I can accept abbreviated variable names, but it is better if abbreviated variable names only appear within a range of 5 lines of code. When abbreviated variable names appear within functions with up to 50 lines, it makes it a little difficult for me to read the code. The abbreviated variable names I referred e.g.: rm []*protobuf.RawMessage, as []*accounts.Account, ss *settings.Settings. I prefer to use rawMessage->rm, account->as, and setting->ss. If 'settings' collides with something, I will use naming such as 'setting' or 'targetSetting' or something else but rather than abbreviated `ss`. 🙂 (I knew you prefer pretty short names rm/as/ss 😉)

Samyoul replied:
That is completely fine. Let me change the variable to `syncSettings`, that way it is explicit and the collision is avoided.

qfrank reviewed (approved):
LGTM! left only minor comments.
